src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/TableParser.cs

```diff
@@ -5,99 +5,158 @@
+                                columnQueues.Remove(colIndex); // once these are all cleared the outer loop will break
+                            }
+                        }
+                        else
+                        {
+                            formattedDataCell = new FormattedCell();
+                        }
+
+                        subsequentRow.Add(formattedDataCell);
+                    }
+
+                    stringTable.Add(subsequentRow);
                 }
             }
 
-            return ToStringTable(values, arrValues);
+            return ToPaddedStringTable(stringTable);
         }
 
-        internal static IEnumerable<string> ToStringTable<T>(this T[] values, string[,] arrValues)
+        internal static IEnumerable<FormattedCell> ToPaddedStringTable(IEnumerable<IEnumerable<FormattedCell>> values)
         {
-            var maxColumnsWidth = GetMaxColumnsWidth(arrValues);
-            var rows = new List<string>();
+            var maxColumnsWidth = GetMaxColumnsWidth(values);
+            var stringTable = new List<FormattedCell>();
 
-            for (var rowIndex = 0; rowIndex < arrValues.GetLength(0); rowIndex++)
+            foreach (var row in values)
             {
-                for (var colIndex = 0; colIndex < arrValues.GetLength(1); colIndex++)
+                int colIndex = 0;
+                foreach (var dataCell in row)
                 {
-                    var cell = arrValues[rowIndex, colIndex];
-                    cell = cell.PadRight(maxColumnsWidth[colIndex]);
-
-                    if (colIndex != 0)
-                    {
-                        cell = "   " + cell;
-                    }
-
-                    rows.Add(cell);
+                    dataCell.Value = "   " + dataCell.Value?.PadRight(maxColumnsWidth[colIndex]) ?? string.Empty;
+                    stringTable.Add(dataCell);
+                    colIndex++;
                 }
 
-                rows.Add(Environment.NewLine);
+                stringTable.Add(new FormattedCell(Environment.NewLine));
             }
 
-            return rows;
+            return stringTable;
         }
 
-        private static int[] GetMaxColumnsWidth(string[,] arrValues)
+        private static int[] GetMaxColumnsWidth(IEnumerable<IEnumerable<FormattedCell>> values)
         {
-            var maxColumnsWidth = new int[arrValues.GetLength(1)];
-            for (var colIndex = 0; colIndex < arrValues.GetLength(1); colIndex++)
+            int[] maxColumnsWidth = null;
+            foreach (var row in values)
             {
-                for (var rowIndex = 0; rowIndex < arrValues.GetLength(0); rowIndex++)
+                // use the first row to get dimension
+                if (maxColumnsWidth == null)
                 {
-                    var newLength = arrValues[rowIndex, colIndex].Length;
-                    var oldLength = maxColumnsWidth[colIndex];
+                    maxColumnsWidth = new int[row.Count()];
```

Comment:
LINQ's `Count()` method is a pet-peeve of mine. It uses reflection to see if the enumeration happens to be a collection that implements a `Count` property, and if not, then it enumerates each value of the enumerable and counts the number of iterations. In other words, while it's "fast enough" in many scenarios, it's still so much slower than using an interface that has a `Count` property built in, which is exactly what `ICollection` and `IReadOnlyCollection` are. Therefore, I suggest you change the input parameter to use one of these interfaces, and get rid of the `()` that makes `Count` unnecesserily slow.  Before joining Microsoft, I worked on a product where data was fetched from a database, and trying to enumerate the enumerable more than once would throw an exception. Doesn't happen for in-memory lists, but it is needlessly slow.

Reply:
Sounds good. Thanks.